cellmlmanip/_singularity_fixes.py

```diff
@@ -0,0 +1,252 @@
+"""
+``remove_fixable_singularities`` specifies a method to remove fixable singularities from a model
+"""
+from math import isclose
+
+import networkx as nx
+from sympy import (
+    Abs,
+    Add,
+    Eq,
+    Float,
+    I,
+    Mul,
+    Piecewise,
+    Pow,
+    Wild,
+    exp,
+    log,
+    solveset,
+)
+
+import cellmlmanip
+from cellmlmanip.model import FLOAT_PRECISION
+
+
+ONE = cellmlmanip.Quantity(1.0, 'dimensionless')
+
+
+def _generate_piecewise(expr, V, sp, Vmin, Vmax):
+    """
+    Generates a new (piecewise) expression based on expr with a linear interpolation when V is between Vmin and Vmax
+
+    The returned expression between Vmin and Vmax is::
+    f(vs) + (f(ve) - f(vs)) * (V - vs) / (ve - vs)
+
+    :param: The expression to turn into a piecewise
+    :param: V the voltage variable.
+    :param: the value of the singularity point: the value of V for whih U == 0
+    :param: the value of the lower bound: the value of V for which U - U_offset == 0
+    :param: the value of the upper bound: the value of V for which U + U_offset == 0
+    :return: expr with a singularity exception around sp, if sp is not None else return the original expr
+    """
+    f_Vmin = expr.xreplace({V: Vmin})
+    f_Vmax = expr.xreplace({V: Vmax})
+    return Piecewise((f_Vmin + ((V - Vmin) / (Vmax - Vmin)) * (f_Vmax - f_Vmin),
+                      Abs(V - sp) < Abs((Vmax - Vmin) / 2)), (expr, True))
+
+
+def _get_singularity(expr, V, U_offset, exp_function):
+    """
+    Finds singularities in equations of the form:
+       - ``U / (exp(U) - 1.0)``
+       - ``U / (1.0 - exp(U))``
+       - ``(exp(U) - 1.0) / U``
+       - ``(1.0 - exp(U)) / U``
+
+    In addition to the singularity itself, this functions returns a lower and upper bounds as follows:
+    - sp is the singularity point: the value of V for whih U == 0
+    - Vmin is the value of V for which U - U_offset == 0
+    - Vmax is the value of V for which U + U_offset == 0:
+    :return: (Vmin, Vmax, singularity point)
+    """
+    Z = Wild('Z', real=True)
+    U_wildcard = Wild('U_wildcard', real=True, include=[V])
+    SP_wildcard = Wild('SP_wildcard', real=True)
+
+    def check_bottom_match(m):
+        return m is not None and U_wildcard in m and Z in m and Z != 0
+
+    def check_top_match(m, sp):
+        return m is not None and Z in m and Z != 0 and (sp == m[SP_wildcard]
+                                                        or (isinstance(sp, Float)
+                                                            and isinstance(m[SP_wildcard], Float)
+                                                            and isclose(m[SP_wildcard], sp)))
+
+    def float_dummies(expr):
```

Comment:
Could these functions-defined-in-a-function become private functions in the module instead?

Reply:
`float_dummies` yes but not the 2 working on the wildcards